### ISSUE TITLE
Question: infinite loop in `l-yaml-stream`?

### ISSUE THREAD

**Issue by yume-chan** [yaml/yaml-spec#166]
`l-yaml-stream` is defined as

```text
[211] l-yaml-stream ::=
  l-document-prefix* l-any-document?
  ( l-document-suffix+ l-document-prefix* l-any-document?
  | l-document-prefix* l-explicit-document? )*
```

Since `l-document-prefix* l-explicit-document?` both can be repeated 0 times, everything (including invalid yaml text) can match this rule. Then since it's in a `()*`, it can match nothing (including eof) infinite times, so it will never report an error or exit after end of file.

Actually, the reference JavaScript parser implementation didn't follow the rule

https://github.com/yaml/yaml-reference-parser/blob/7a5c87e5dd0586db814fb928a3e45fdbdb31216a/javascript/lib/grammar.js#L1122-L1125

```js
this.all(
  this.l_document_prefix,
  this.rep(0, 1, this.l_any_document),
  this.rep2(
    0,
    null,
    this.any(
      this.all(
        this.l_document_suffix,
        this.rep(0, null, this.l_document_prefix),
        this.rep2(0, 1, this.l_any_document)
      ),
      this.all(this.l_document_prefix, this.rep(0, 1, this.l_explicit_document))
    )
  )
);
```

(it didn't repeat `l-document-prefix` before `l-explicit-document`)

**Comment by yume-chan** [comment]
`[202]	l-document-prefix	::=	c-byte-order-mark? l-comment*`

`l-document-prefix` itself can match nothing, so `l-document-prefix*` will also cause infinite loop.

**Comment by ingydotnet** [comment]
Hi @yume-chan,

The reference parser project uses this file
https://github.com/yaml/yaml-grammar/blob/master/yaml-spec-1.2.yaml
which is programmatically generated from the 1.2 spec HTML.

Then is applies this patch file to fix all the bugs uncovered in the 1.2 productions.
https://github.com/yaml/yaml-reference-parser/blob/main/build/yaml-spec-1.2.patch#L319-L352

The code you referenced is generated from the patched grammar.

As you can see the patch fixes 211.

This will also be fixed in the YAML 1.2.2 spec which will be published very soon.

**Comment by yume-chan** [comment]
https://github.com/yaml/yaml-reference-parser/blob/7a5c87e5dd0586db814fb928a3e45fdbdb31216a/build/yaml-spec-1.2.patch#L325-L329

```text
+# l-yaml-stream ::=
+#   l-document-prefix l-any-document?
+#   ( ( l-document-suffix+ l-document-prefix
+#   l-any-document? )
+#   | ( l-document-prefix l-explicit-document? ) )*
```

I don't think this fix the problem. `l-document-prefix` can match nothing, `l-explicit-document?` can match nothing, `( l-document-prefix l-explicit-document? )*` can still match nothing infinite times.

**Comment by ingydotnet** [comment]
Recalling this work from last year, the bug was that `l-document-prefix` could match multiple BOMs which wasn't intended.

I'm not a fan of #211 but I think the patched version is right.

Parsing stops when you reach the end of a stream having matched the starting production (211 in this case).

Consider the regex matching operation:
```
"foo".match(/foox*/)
```
this matches. It doesn't try to match `x` infinite times.

If you know a better way to express 211 (without changing other productions) let me know. I'm actually refactoring the productions for the imminent 1.2.2 spec revision.

**Comment by yume-chan** [comment]
The regex example hinted me where the problem is. Regex and YAML JavaScript reference parser use different parsing methods. 

Regex read each character from input, and compare with rules, so it knows when the input ends and stop.

YAML JavaScript reference parser and my parsers work on plain strings, let each rules check the input, and chooses a succeed one. They don't know (or care) whether the input is ended. It's a problem in the parser, not the spec.

In YAML JavaScript reference parser, repeated rules must consume at least one character from the input (https://github.com/yaml/yaml-reference-parser/blob/7a5c87e5dd0586db814fb928a3e45fdbdb31216a/javascript/lib/parser.js#L251-L253). This may hide infinite loops in the middle of input, but if all rules are defined and implemented correctly, that shouldn't happen.

**Comment by ingydotnet** [comment]
704b02c was just merged to make this not ambiguous.

The production still has problems but they can't be addressed until 1.3.
